tests/components/roku/test_config_flow.py

```diff
@@ -0,0 +1,247 @@
+        new=MockDeviceInfo,
+    ):
+        return await hass.config_entries.flow.async_configure(
+            flow_id=flow_id, user_input=user_input
+        )
+
+
+async def async_init_flow(
+    hass: HomeAssistantType,
+    handler: str = DOMAIN,
+    context: Optional[Dict] = None,
+    data: Any = None,
+) -> Any:
+    """Set up mock Roku integration flow."""
+    with patch(
+        "homeassistant.components.roku.config_flow.Roku.device_info",
+        new=MockDeviceInfo,
+    ):
+        return await hass.config_entries.flow.async_init(
+            handler=handler, context=context, data=data
+        )
+
+
+async def test_duplicate_error(hass: HomeAssistantType) -> None:
+    """Test that errors are shown when duplicates are added."""
+    await setup_integration(hass, skip_entry_setup=True)
+
+    result = await async_init_flow(
+        hass, context={CONF_SOURCE: SOURCE_IMPORT}, data={CONF_HOST: HOST}
+    )
+
+    assert result["type"] == RESULT_TYPE_ABORT
+    assert result["reason"] == "already_configured"
+
+    result = await async_init_flow(
+        hass, context={CONF_SOURCE: SOURCE_USER}, data={CONF_HOST: HOST}
+    )
+
+    assert result["type"] == RESULT_TYPE_ABORT
+    assert result["reason"] == "already_configured"
+
+    result = await async_init_flow(
+        hass,
+        context={CONF_SOURCE: SOURCE_SSDP},
+        data={
+            ATTR_UPNP_FRIENDLY_NAME: UPNP_FRIENDLY_NAME,
+            ATTR_SSDP_LOCATION: SSDP_LOCATION,
+            ATTR_UPNP_SERIAL: UPNP_SERIAL,
+        },
+    )
+
+    assert result["type"] == RESULT_TYPE_ABORT
+    assert result["reason"] == "already_configured"
+
+
+async def test_form(hass: HomeAssistantType) -> None:
+    """Test the user step."""
+    await async_setup_component(hass, "persistent_notification", {})
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={CONF_SOURCE: SOURCE_USER}
+    )
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["errors"] == {}
+
+    with patch(
+        "homeassistant.components.roku.async_setup", return_value=True
+    ) as mock_setup, patch(
+        "homeassistant.components.roku.async_setup_entry", return_value=True,
+    ) as mock_setup_entry:
+        result = await async_configure_flow(hass, result["flow_id"], {CONF_HOST: HOST})
+
+    assert result["type"] == RESULT_TYPE_CREATE_ENTRY
+    assert result["title"] == HOST
+    assert result["data"] == {CONF_HOST: HOST}
+    await hass.async_block_till_done()
+    assert len(mock_setup.mock_calls) == 1
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_cannot_connect(hass: HomeAssistantType) -> None:
+    """Test we handle cannot connect roku error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={CONF_SOURCE: SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.roku.config_flow.validate_input",
```

Comment:
We want to patch the device library class method, not our code.

Reply:
WIP https://github.com/home-assistant/core/pull/32883